test/distributed/_tensor/test_random_ops.py

```diff
@@ -122,6 +119,91 @@ def test_deterministic_dropout_1d(self):
                         local_tensor_list[self.rank], local_tensor_list[other_rank]
                     )
 
+    @with_comms
+    @skip_unless_torch_gpu
+    def test_deterministic_uniform_nd(self):
+        mesh = torch.arange(self.world_size).reshape(2, 2, -1)
+        device_mesh = DeviceMesh(self.device_type, mesh)
+        # using world_size guarantees that each dim can be evenly partitioned
+        dtensor_size = [self.world_size for l in mesh.size()]
+        # initialize rng state
+        manual_seed(1234, device_mesh)
+        self.check_rng_state(1234, 0, device_mesh)
+
+        placements_list = [  # this list of placements should be enough to cover
+            [Shard(0), Shard(1), Shard(2)],
+            [Shard(2), Shard(1), Shard(0)],
+            [Shard(1), Replicate(), Shard(0)],
+            [Replicate(), Replicate(), Replicate()],
+        ]
+        dim_map_list = [
+            [0, 1, 2],
+            [2, 1, 0],
+            [2, 0, -1],
+            [-1, -1, -1],
+        ]
+
+        coord = device_mesh.get_coordinate()
+        assert coord is not None
+
+        for placements, dim_map in zip(placements_list, dim_map_list):
+            # shard shape:
+            shard_shape = [mesh.size()[dim] if dim >= 0 else 1 for dim in dim_map]
+            # shard coord:
+            shard_coord = [coord[dim] if dim >= 0 else 0 for dim in dim_map]
+            strides = [1]
+            for l in shard_shape[:0:-1]:
+                strides.append(strides[-1] * l)
+            strides = strides[::-1]
```

Comment:
why we need to calculate shape and strides in the tests? I thought we only need to calculate them inside `random.py`?

Reply:
The reason is:
1) I want to check that all shards are different, like what I did in the `1d` test case.
2) For different ranks, they may have the same shard. example: `mesh = [[0, 1], [2, 3]], placements = [Shard(0), Replicate()]`. Rank 0 and 1 have the same shard. So I need to compute which shard each rank actually has to determine if it should be a `self.assertEqual` or  `self.assertNotEqual`. 

An improvement is to extract this logic as an internal API and call it in test code as well as `set_pre_op_offset`.